File: LeetCode203/LeetCode203.cpp
// Remove all elements from a linked list of integers that have value val.
// Example:
// Input:  1->2->6->3->4->5->6, val = 6
// Output: 1->2->3->4->5

#include <iostream>

//  Definition for singly-linked list.
// struct ListNode
// {
//     int val;
//     ListNode *next;
//     ListNode(int x) : val(x), next(NULL) {}
// };

class Solution
{
public:
    ListNode *removeElements(ListNode *head, int val)
    {

        while (head != nullptr && head->val == val)
            head = head->next;
        if (head == nullptr)
            return head;

        ListNode *pointer = head;
        ListNode *freePointer = nullptr;

        while (pointer->next != nullptr)
        {
            std::cout << pointer->val << " ";

            if (pointer->next->val == val)
            {
                freePointer = pointer->next;
                pointer->next = freePointer->next;
                delete (freePointer);
            }
            else if (pointer->next != nullptr)
                pointer = pointer->next;
        }

        return head;
    }
};

File: LeetCode203/LeetCode203.java
// Remove all elements from a linked list of integers that have value val.
// Example:
// Input:  1->2->6->3->4->5->6, val = 6
// Output: 1->2->3->4->5

// Definition for singly-linked list.
// public class ListNode {
//     int val;
//     ListNode next;

//     ListNode(int x) {
//         val = x;
//     }
// }

class Solution {
    public ListNode removeElements(ListNode head, int val) {
        ListNode prehead = new ListNode(-1);
        prehead.next = head;
        ListNode pointer = prehead;

        while (pointer != null && pointer.next != null)
            if (pointer.next.val == val)
                pointer.next = pointer.next.next;
            else
                pointer = pointer.next;

        return prehead.next;
    }
}
